### PR TITLE
Kore Manual

### PR DESCRIPTION
**Pull request by dopamane**
Fixes https://github.com/kframework/kore/issues/2806

For an explanation of the document structure, section meanings, and topic ideas, see https://github.com/kframework/kore/pull/2845#issuecomment-926295641

Thoughts on hooks: https://github.com/kframework/kore/pull/2845#issuecomment-926281929

Thoughts on profiling: https://github.com/kframework/kore/pull/2845#issuecomment-926319803

Thoughts on improved section numbering: https://github.com/kframework/kore/pull/2845#issuecomment-927452037

Thoughts on adding builtin hooks info: https://github.com/kframework/kore/pull/2845#discussion_r715497574

Existing documentation included in this PR:

* [Glossary Wiki page](https://github.com/kframework/kore/wiki/Glossary)
* [hooks.md partial (only Bool, Int), still needs String, Map, List, Set, KEqual, Krypto, and Bytes](https://github.com/kframework/kore/blob/master/docs/hooks.md)
* [Google drive context map images and explanations](https://drive.google.com/drive/folders/1EGkXyayfWWMB5miSL9EfHjjCy-bqiqbJ)
* [Builtin symbols Wiki page](https://github.com/kframework/kore/wiki/Builtin-symbols)
* [kore-syntax.md](https://github.com/kframework/kore/blob/master/docs/kore-syntax.md)
* [kore-implicits.md](https://github.com/kframework/kore/blob/master/docs/kore-implicits.md)

For example of inline math notation, see glossary "substitution", "function-like", and "functional".
 
---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

**Comment by dopamane** [comment]
I think the list of hooks might need to be moved to an appendix, leaving the hooks section to describe how to add new hooks and their Haskell implementation. But I left it in Implementation section for now until we see more documentation populated around it.

**Comment by dopamane** [comment]
Section explanation:
* Introduction: Why is Kore important? What does it let us do? Show the big picture! It's an awesome project, we can use this section to make a good impression and motivate the reader to explore more. Also discuss the structure of the rest of the document.
* Basics: To make this document self-contained we need to provide the reader with tools to understand later sections like Design and Implementation. We can explain general ideas like symbolic execution, basic matching logic, unification etc. assuming no prior knowledge.
* Design: This is the idea of Kore, our blueprint as developers. It will probably be a math/logic-heavy section. We can describe the mental models necessary to understand how Kore works. We should use ample diagrams and perhaps even GIFs to illustrate high-level algorithms. This should be complete and accessible without needing knowledge of Haskell or examination of code.
* Implementation: Haskell heavy section. Document data types, functions, algorithms with code examples. We can draw from existing haddocks. This includes build system information and source structure. We also have context maps to describe module structure and dependencies. We should document utilities like logging here as well.
* Language: Complete description of Kore language including syntax, lexical structure, and validity.
* Commands: Reference command line tools, demonstrate usage.
* K Framework: How does Kore fit into the larger K Framework, what are the interfaces it has with the K project.
* Application Notes: Any relevant papers or publications can be reproduced here or referenced.
* Glossary: define vocabulary used throughout this document
* Appendix: We can place large reference tables or lists of hooks here if necessary. Bulky items that breakup the flow of the document should live here.

**Comment by dopamane** [comment]
I think documentation concerning Haskell profiling can live under Implementation because it is most relevant to Kore developers. When we are sure it can stand alone, perhaps we add a Profiling section.

File: docs/manual/DEVELOPER_MANUAL.md [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+
+### [Hooks]{#implementation-hooks}
+
+[Pull request #632](https://github.com/kframework/kore/pull/632/files) is a self-contained example of adding a builtin symbol. When adding a builtin symbol we should add:
+
+1. An entry to the hooks document.
+2. A symbol verifier.
+3. An evaluator for concrete arguments (at least).
+4. A declaration to `Test.Kore.Builtin.Definition`.
+5. A few tests with different concrete arguments.
```
Comment (ana-pantilie) [inline docs/manual/DEVELOPER_MANUAL.md:200]: Let's keep this for now, but remember it's probably very out of date (not necessarily the method itself, but the PR we are linking to). We should probably add a TODO here or something to remind us.

**Review by ana-pantilie — COMMENTED** [review]
Can you link to your comment describing the way you envisioned the sections in the PR description? They're great ideas, and I almost missed it.

Also, please point out which documents/wiki entries you added to the manual in this PR.

One more question: where is the sample LaTeX? I couldn't find it, but maybe I skipped over it by accident.

**Review by ana-pantilie — APPROVED** [review]

File: docs/manual/DEVELOPER_MANUAL.md (hunk 1 above)
Comment (dopamane) [reply docs/manual/DEVELOPER_MANUAL.md]: Added TODO with link to your comment, also made this information it's own section for easy access via table of contents.

**Comment by dopamane** [comment]
We should also number the sections and table of contents better to make it easy for external reference, E.g. Language - Syntax - Sentences can be written as 5.2.4 . So table of contents needs work, numbered lists won't cut it. I think just using `*` and making our own numbering is better:

```
* 5 Language
    * 5.1 Lexical Structure
        * 5.1.1 Comments
        * 5.1.2 String literals
        * 5.1.3 Keywords
        * 5.1.4 Identifiers
    * 5.2 Syntax
        * 5.2.1 Sorts
        * 5.2.4 Sentences
```

**Pull request merged**